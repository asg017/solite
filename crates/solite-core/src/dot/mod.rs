mod bench;
mod load;
mod export;
mod open;
mod param;
mod print;
mod sh;
mod tables;
mod timer;
mod vegalite;

pub use crate::dot::{
  bench::BenchCommand,
  load::LoadCommand,
  sh::ShellCommand,
  print::PrintCommand,
  tables::TablesCommand,
  open::OpenCommand,
  export::ExportCommand,
  vegalite::VegaLiteCommand,
  param::ParameterCommand,
};
pub use load::LoadCommandSource;

use param::parse_parameter;
use thiserror::Error;

use crate::Runtime;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Error, Debug, PartialEq)]
pub enum ParseDotError {
    #[error("Unknown command '{0}'")]
    UnknownCommand(String),
    #[error("Invalid argument: {0}")]
    InvalidArgument(String),

    #[error("{0}")]
    Generic(String),
}





#[derive(Serialize, Debug)]
pub enum DotCommand {
    /*  introspection  */
    //Databases,
    Tables(TablesCommand),
    //Indexes,
    //Schema,

    /*  docs  */
    //Help,
    //Docs,

    /*  runtime  */
    /// .run file.sql -p a=1 -p b=2
    //Run,
    /// or .debugger?
    //Breakpoint,

    //Quit,
    //Exit,

    /// switches to different DB connection
    /// usage: .open file.db
    Open(OpenCommand),
    Load(LoadCommand),

    /// TODO sqlpkg/spm support
    //Load,
    Print(PrintCommand),

    Shell(ShellCommand),
    /// usage: .param set name 'alex garcia'
    Parameter(ParameterCommand),
    /// usage: .bail on/off
    //Bail,

    /// usage: .timer on/off
    Timer(bool),
    //
    //Mode,
    Export(ExportCommand),
    Vegalite(VegaLiteCommand),
    Bench(BenchCommand),
}

fn parse_bool(s: String) -> Result<bool, String> {
    match s.to_lowercase().as_str() {
        "yes" | "y" | "on" => Ok(true),
        "no" | "n" | "off" => Ok(false),
        _ => Err(format!("Not a boolean value: {}", s)),
    }
}
pub fn parse_dot<S: Into<String>>(
    command: S,
    args: S,
    rest: &str,
    runtime: &mut Runtime,
) -> Result<DotCommand, ParseDotError> {
    let command = command.into();
    let args = args.into();
    match command.to_lowercase().as_str() {
        "print" => Ok(DotCommand::Print(PrintCommand { message: args })),
        "sh" => Ok(DotCommand::Shell(ShellCommand { command: args })),
        "tables" => Ok(DotCommand::Tables(TablesCommand {})),
        "open" => Ok(DotCommand::Open(OpenCommand { path: args })),
        "export" => Ok(DotCommand::Export(ExportCommand::new(args, runtime, rest)?)),
        "bench" => Ok(DotCommand::Bench(BenchCommand::new(args, runtime, rest)?)),
        "vl" | "vegalite" => Ok(DotCommand::Vegalite(VegaLiteCommand::new(
            args, runtime, rest,
        )?)),
        "load" => Ok(DotCommand::Load(LoadCommand::new(args))),
        "timer" => Ok(DotCommand::Timer(
            parse_bool(args).map_err(ParseDotError::InvalidArgument)?,
        )),
        "param" | "parameter" => Ok(DotCommand::Parameter(parse_parameter(args))),
        _ => Err(ParseDotError::UnknownCommand(command)),
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_parse_dot() {}
}
