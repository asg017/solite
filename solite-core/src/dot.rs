use solite_stdlib::solite_stdlib_init;
use thiserror::Error;

use crate::{sqlite::Connection, Runtime};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Error, Debug, PartialEq)]
pub enum ParseDotError {
    #[error("Unknown command '{0}'")]
    UnknownCommand(String),
    #[error("Invalid argument: {0}")]
    InvalidArgument(String),
}

#[derive(Serialize, Debug, PartialEq)]
pub struct PrintCommand {
    pub message: String,
}
impl PrintCommand {
    pub fn execute(&self) {
        println!("{}", self.message);
    }
}

#[derive(Serialize, Debug, PartialEq)]
pub struct TablesCommand {}
impl TablesCommand {
    pub fn execute(&self, runtime: &Runtime) {
        let stmt = runtime
            .connection
            .prepare(
                r#"
                select name
                from pragma_table_list
                where "schema" = 'main'
                  and type in ('table', 'view')
                  and name not like 'sqlite_%'
                order by name;
                "#,
            )
            .unwrap()
            .1
            .unwrap();
        let mut tables = vec![];
        while let Ok(Some(row)) = stmt.next() {
            tables.push(row.get(0).unwrap().as_str().to_owned());
        }
        for table in tables {
            println!("{table}")
        }
    }
}

#[derive(Serialize, Debug, PartialEq)]
pub struct OpenCommand {
    pub path: String,
}
impl OpenCommand {
    pub fn execute(&self, runtime: &mut Runtime) {
        runtime.connection = Connection::open(&self.path).unwrap();
        unsafe {
          solite_stdlib_init(runtime.connection.db(), std::ptr::null_mut(), std::ptr::null_mut());
        }
    }
}

#[derive(Serialize, Debug, PartialEq)]
pub struct LoadCommand {
    pub path: String,
    pub entrypoint: Option<String>,
}
impl LoadCommand {
    pub fn new(args: String) -> Self {
        let (path, entrypoint) = match args.split_once(' ') {
            Some((path, entrypoint)) => (path.to_string(), Some(entrypoint.trim().to_string())),
            None => (args, None),
        };
        Self { path, entrypoint }
    }
    pub fn execute(&self, connection: &mut Connection) {
        connection.load_extension(&self.path, &self.entrypoint);
    }
}

#[derive(Serialize, Debug, PartialEq)]
pub enum ParameterCommand {
    Set { key: String, value: String },
    Unset(String),
    List,
    Clear,
}

fn parse_parameter(line: String) -> ParameterCommand {
    match line.trim_end().split_once(' ') {
        Some((word, rest)) => match word {
            "set" => {
                let (k, v) = rest.split_once(' ').unwrap();
                ParameterCommand::Set {
                    key: k.to_owned(),
                    value: v.to_owned(),
                }
            }
            "unset" => ParameterCommand::Unset(rest.to_owned()),
            _ => todo!(),
        },
        None => match line.trim_end() {
            "list" => ParameterCommand::List,
            "clear" => ParameterCommand::Clear,
            _ => todo!(),
        },
    }
}
#[derive(Serialize, Debug, PartialEq)]
pub enum DotCommand {
    /*  introspection  */
    //Databases,
    Tables(TablesCommand),
    //Indexes,
    //Schema,

    /*  docs  */
    //Help,
    //Docs,

    /*  runtime  */
    /// .run file.sql -p a=1 -p b=2
    //Run,
    /// or .debugger?
    //Breakpoint,

    //Quit,
    //Exit,

    /// switches to different DB connection
    /// usage: .open file.db
    Open(OpenCommand),
    Load(LoadCommand),

    /// TODO sqlpkg/spm support
    //Load,
    Print(PrintCommand),

    /// usage: .param set name 'alex garcia'
    Parameter(ParameterCommand),
    /// usage: .bail on/off
    //Bail,

    /// usage: .timer on/off
    Timer(bool),
    //
    //Mode,
    //Export,
}

fn parse_bool(s: String) -> Result<bool, String> {
    match s.to_lowercase().as_str() {
        "yes" | "y" | "on" => Ok(true),
        "no" | "n" | "off" => Ok(false),
        _ => Err(format!("Not a boolean value: {}", s)),
    }
}
pub fn parse_dot<S: Into<String>>(command: S, args: S) -> Result<DotCommand, ParseDotError> {
    let command = command.into();
    let args = args.into();
    match command.to_lowercase().as_str() {
        "print" => Ok(DotCommand::Print(PrintCommand { message: args })),
        "tables" => Ok(DotCommand::Tables(TablesCommand {})),
        "open" => Ok(DotCommand::Open(OpenCommand { path: args })),
        "load" => Ok(DotCommand::Load(LoadCommand::new(args))),
        "timer" => Ok(DotCommand::Timer(
            parse_bool(args).map_err(ParseDotError::InvalidArgument)?,
        )),
        "param" | "parameter" => Ok(DotCommand::Parameter(parse_parameter(args))),
        _ => Err(ParseDotError::UnknownCommand(command)),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_dot() {
        assert_eq!(
            parse_dot("print", "asdf"),
            Ok(DotCommand::Print(PrintCommand {
                message: "asdf".to_owned()
            }))
        );
        assert_eq!(
            parse_dot("unknown", ""),
            Err(ParseDotError::UnknownCommand("unknown".to_string()))
        );
    }
}
