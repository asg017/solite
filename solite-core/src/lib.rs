pub mod dot;
pub mod replacement_scans;
pub mod sqlite;

use crate::sqlite::Connection;
use dot::{parse_dot, ParseDotError};
use libsqlite3_sys::{
    sqlite3_db_config, SQLITE_DBCONFIG_DEFENSIVE, SQLITE_DBCONFIG_WRITABLE_SCHEMA,
};
use ropey::Rope;
use solite_stdlib::solite_stdlib_init;
use sqlite::{OwnedValue, SQLiteError, Statement};
use std::{fmt, fmt::Write, path::PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum StepError {
    #[error("Error preparing SQL statement:")]
    Prepare {
        file_name: String,
        src: String,
        offset: usize,
        error: SQLiteError,
    },
    #[error("Error parsing dot command: {0}")]
    ParseDot(ParseDotError),
}

/// A "block" of commands to run - can be SQL, PRQL(?) or dot commands.
/// Can come from a SQL file ondisk, a Jupyer cell, REPL, etc.
#[derive(Debug)]
pub struct Block {
    // either file name or "[stdin]" or "[script]" or something
    name: String,
    source: BlockSource,
    contents: String,
    rope: Rope,
    offset: usize,
}

#[derive(Debug)]
pub enum BlockSource {
    File(PathBuf),
    Repl,
    JupyerCell,
}

#[derive(Default)]
pub struct State {
    //timer: bool,
    //bail: bool,
}

pub struct Runtime {
    pub connection: Connection,
    stack: Vec<Block>,
    //state: State,
    initialized_sqlite_parameters_table: bool,
}

pub struct StepReference {
  block_name: String,
  line_number: usize,
  column_number: usize,
}

impl fmt::Display for StepReference {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
      write!(f, "{}:{}:{}", self.block_name, self.line_number, self.column_number)
  }
}

pub enum StepResult {
    SqlStatement{stmt: Statement, raw_sql: String},
    DotCommand(dot::DotCommand),
}
pub struct Step {
    /// Dot command or SQL
    pub result: StepResult,


    pub reference: StepReference,
}

fn error_context(error: &SQLiteError, block: &Block) -> Result<String, std::fmt::Error> {
    let mut ctx = String::new();

    let error_offset = error.offset.unwrap_or(0) + block.offset;
    let line_idx_with_error = block.rope.byte_to_line(error_offset);
    let line_error_offset = error_offset - block.rope.line_to_byte(line_idx_with_error);
    let longest_digit_length =
        ((line_idx_with_error + 1 + 4).checked_ilog10().unwrap_or(0) + 1) as usize;
    writeln!(
        &mut ctx,
        "{} ({})",
        error.code_description, error.result_code
    )?;

    for b in (1..4).rev() {
        if let Some(idx) = line_idx_with_error.checked_sub(b) {
            if let Some(line) = block.rope.get_line(idx) {
                writeln!(
                    &mut ctx,
                    "{:longest_digit_length$} | {}",
                    line_idx_with_error + 1 - b,
                    line.as_str().unwrap().trim_end_matches('\n'),
                )?;
            }
        };
    }
    writeln!(
        &mut ctx,
        "{:longest_digit_length$} | {}",
        line_idx_with_error + 1,
        block
            .rope
            .line(line_idx_with_error)
            .as_str()
            .unwrap()
            .trim_end_matches('\n'),
    )?;
    writeln!(
        &mut ctx,
        "{:longest_digit_length$} | {}^ {}",
        " ",
        " ".repeat(line_error_offset),
        error.message,
    )?;
    for b in 1..4 {
        if let Some(line) = block.rope.get_line(line_idx_with_error + b) {
            writeln!(
                &mut ctx,
                "{:longest_digit_length$} | {}",
                line_idx_with_error + 1 + b,
                line.as_str().unwrap().trim_end_matches('\n'),
            )?;
        }
    }

    writeln!(
        &mut ctx,
        "\tat {:longest_digit_length$}:{}:{}",
        match &block.source {
            BlockSource::File(p) => p.to_string_lossy().to_string(),
            BlockSource::JupyerCell => "[cell]".to_owned(),
            BlockSource::Repl => "[repl]".to_owned(),
        },
        line_idx_with_error + 1,
        line_error_offset + 1,
    )?;

    Ok(ctx)
}

impl Runtime {
    pub fn new(path: Option<String>) -> Self {
        unsafe {
            libsqlite3_sys::sqlite3_auto_extension(Some(std::mem::transmute(
                solite_stdlib_init as *const (),
            )));
        }
        let connection = match path {
            Some(path) => Connection::open(path.as_str()).unwrap(),
            None => Connection::open_in_memory().unwrap(),
        };
        Runtime {
            connection,
            stack: vec![],
            //state: State::default(),
            initialized_sqlite_parameters_table: false,
        }
    }
    pub fn enqueue(&mut self, name: &str, code: &str, source: BlockSource) {
        self.stack.push(Block {
            name: name.to_string(),
            source,
            contents: code.to_string(),
            rope: Rope::from_str(code),
            offset: 0,
        });
    }
    
    // temporary for snapshot
    pub fn next_sql_step(&mut self) -> Result<Option<Step>, StepError> {
      let mut current = match self.stack.pop() {
        Some(x) => x,
        None => return Ok(None),
      };
      let full = (current.contents).get(current.offset..).unwrap();
      let code = full.to_owned();
      let local_offset = 0;
      match self.prepare_with_parameters(&code) {
        Ok((rest, Some(stmt))) => {
            let stmt_offset_idx = if (current.offset - local_offset) == 0 {
                0
            } else {
                current.offset - local_offset + 1
            };
            let line_idx = current.rope.byte_to_line(stmt_offset_idx);
            let column_idx = stmt_offset_idx - current.rope.line_to_byte(line_idx);
            let block_name = current.name.clone();
            if let Some(rest) = rest {
                current.offset += rest;
                self.stack.insert(0, current);
            }
            return Ok(Some(Step {
                reference: StepReference {
                  block_name,
                  line_number: line_idx + 1,
                  column_number: column_idx + 1,
                },
                result: StepResult::SqlStatement{stmt, raw_sql: code.to_owned()},
            }));
        }
        Ok((_rest, None)) => {
            return Ok(None)
        }
        Err(error) => todo!(),
    }
    }
    pub fn next_step(&mut self) -> Result<Option<Step>, StepError> {
        // loop here handles:
        // 1. Dot commands
        // 2. Running SQL statements
        // 3. Potential replacement scans
        while let Some(mut current) = self.stack.pop() {
            // "code" that starts with whitespace/comments makes it hard to parse
            // out dot commands correctly. So we advance the offset until we reach
            // the first "real" code.
            //let code = (current.contents).get(current.offset..).unwrap();
            let full = (current.contents).get(current.offset..).unwrap();
            let code = advance_through_ignorable(full).to_owned();
            let local_offset = full.len() - code.len();
            current.offset += local_offset;
            if code.starts_with('.') {
                let end_idx = code.find('\n').unwrap_or(code.len());
                let dot_line = code.get(0..end_idx).unwrap();
                let sep_idx = code.find(' ').unwrap_or(code.len());
                let dot_command = dot_line.get(1..sep_idx).unwrap().trim().to_string();
                let dot_args = dot_line.get(sep_idx..).unwrap().trim().to_string();
                let rest = code.get(end_idx..).unwrap();
                let source = current.name.to_string();
                if !rest.is_empty() {
                    current.offset += end_idx + 1;
                    self.stack.push(current);
                }
                let cmd = parse_dot(dot_command, dot_args).map_err(StepError::ParseDot)?;
                return Ok(Some(Step {
                    reference: StepReference {
                      block_name: source,
                      // TODO: why hardcode here?
                      line_number: 1,
                      column_number: 1,
                    },
                    result: StepResult::DotCommand(cmd),
                }));
            } else {
                match self.prepare_with_parameters(&code) {
                    Ok((rest, Some(stmt))) => {
                        let stmt_offset_idx = if (current.offset - local_offset) == 0 {
                            0
                        } else {
                            current.offset - local_offset + 1
                        };
                        let line_idx = current.rope.byte_to_line(stmt_offset_idx);
                        let column_idx = stmt_offset_idx - current.rope.line_to_byte(line_idx);
                        let block_name = current.name.clone();
                        if let Some(rest) = rest {
                            current.offset += rest;
                            self.stack.insert(0, current);
                        }
                        return Ok(Some(Step {
                            reference: StepReference {
                              block_name,
                              line_number: line_idx + 1,
                              column_number: column_idx + 1,
                            },
                            result: StepResult::SqlStatement{stmt, raw_sql: code.to_owned()},
                        }));
                    }
                    Ok((_rest, None)) => {
                        continue;
                    }
                    Err(error) => {
                        match replacement_scans::replacement_scan(&error, &self.connection) {
                            Some(Ok(stmt)) => {
                                stmt.execute().unwrap();
                                self.stack.push(current);
                            }
                            Some(Err(_)) => todo!(),
                            None => {
                                let context = error_context(&error, &current)
                                    .map_err(|_e| "error_context error???".to_owned())
                                    .unwrap();
                                return Err(StepError::Prepare {
                                    error,
                                    file_name: current.name,
                                    src: current.contents,
                                    offset: current.offset,
                                });
                            }
                        };
                    }
                }
            }
        }

        Ok(None)
    }

    pub fn execute_to_completion(&mut self) -> Result<(), StepError> {
        loop {
            match self.next_step() {
                Ok(None) => return Ok(()),
                Ok(Some(step)) => {
                    match step.result {
                        StepResult::SqlStatement{stmt, ..} => stmt.execute().unwrap(),
                        StepResult::DotCommand(_cmd) => todo!(),
                    };
                    continue;
                }
                Err(err) => return Err(err),
            }
        }
    }
    pub fn has_next(&self) -> bool {
        !self.stack.is_empty()
    }

    fn init_sqlite_parameters_table(&mut self) {
        if self.initialized_sqlite_parameters_table {
            return;
        }
        unsafe {
            let original_writable = 0;
            let original_defensive = 0;
            let db = self.connection.db();
            sqlite3_db_config(db, SQLITE_DBCONFIG_DEFENSIVE, -1, &original_defensive);
            sqlite3_db_config(db, SQLITE_DBCONFIG_DEFENSIVE, 0, 0);
            sqlite3_db_config(db, SQLITE_DBCONFIG_WRITABLE_SCHEMA, -1, &original_writable);
            sqlite3_db_config(db, SQLITE_DBCONFIG_WRITABLE_SCHEMA, 1, 0);
            let result = self.connection.prepare("CREATE TABLE IF NOT EXISTS temp.sqlite_parameters(key TEXT PRIMARY KEY, value) WITHOUT ROWID");
            result.unwrap().1.expect("TODO").execute().unwrap();
            sqlite3_db_config(db, SQLITE_DBCONFIG_DEFENSIVE, original_defensive, 0);
            sqlite3_db_config(db, SQLITE_DBCONFIG_WRITABLE_SCHEMA, original_writable, 0);
        }
        self.initialized_sqlite_parameters_table = true;
        //self.connection.prepare("CREATE TABLE TEMP.sqlite_parameters(key text, value any ) WITHOUT ROWID")
    }

    pub fn define_parameter(&mut self, key: String, value: String) -> Result<(), String> {
        self.init_sqlite_parameters_table();
        let stmt = self
            .connection
            .prepare("INSERT OR REPLACE INTO temp.sqlite_parameters(key, value) VALUES (?1, ?2)")
            .unwrap()
            .1
            .unwrap();
        stmt.bind_text(1, key);
        stmt.bind_text(2, value);
        stmt.execute().unwrap();
        Ok(())
    }

    pub fn prepare_with_parameters(
        &self,
        sql: &str,
    ) -> Result<(Option<usize>, Option<Statement>), SQLiteError> {
        let (rest, stmt) = self.connection.prepare(sql)?;
        if let Some(stmt) = stmt {
            let params = stmt.bind_parameters();
            for (idx, param) in params.iter().enumerate() {
                let param = if let Some(param) = param.strip_prefix(':') {
                    param
                } else if let Some(param) = param.strip_prefix('@') {
                    param
                } else {
                    param
                };
                match self.lookup_parameter(param) {
                    Some(OwnedValue::Text(s)) => {
                        stmt.bind_text((idx + 1) as i32, std::str::from_utf8(&s).unwrap())
                    }
                    Some(OwnedValue::Integer(v)) => stmt.bind_int64((idx + 1) as i32, v),
                    Some(OwnedValue::Double(v)) => stmt.bind_double((idx + 1) as i32, v),
                    Some(OwnedValue::Blob(v)) => stmt.bind_blob((idx + 1) as i32, v.as_ref()),
                    Some(OwnedValue::Null) => stmt.bind_null((idx + 1) as i32),

                    None => (),
                }
            }
            Ok((rest, Some(stmt)))
        } else {
            Ok((rest, stmt))
        }
    }
    fn lookup_parameter<S: AsRef<str>>(&self, key: S) -> Option<OwnedValue> {
        let stmt = self
            .connection
            .prepare("SELECT value FROM temp.sqlite_parameters WHERE key = ?1")
            .ok()?
            .1?;
        stmt.bind_text(1, key);
        stmt.next()
            .unwrap()
            .map(|v| OwnedValue::from_value_ref(v.get(0).unwrap()))
    }
}

pub fn advance_through_ignorable(contents: &str) -> &str {
    let mut chars = contents.char_indices();
    let mut latest = 0;

    while let Some((idx, c)) = chars.next() {
        latest = idx;
        if c.is_whitespace() {
            continue;
        }
        if c == '-' {
            if let Some((_, '-')) = chars.next() {
                loop {
                    match chars.next() {
                        Some((idx, '\n')) => {
                            latest = idx;
                            break;
                        }
                        Some((_, _)) => continue,
                        None => break,
                    }
                }
            } else {
                break;
            }
        } else if c == '/' {
            if let Some((_idx, '*')) = chars.next() {
                loop {
                    match chars.next() {
                        Some((idx, '*')) => match chars.next() {
                            Some((idx, '/')) => {
                                latest = idx;
                                break;
                            }
                            Some((idx, _)) => {
                                latest = idx;
                                continue;
                            }
                            None => {
                                latest = idx;
                                break;
                            }
                        },
                        Some((idx, _)) => {
                            latest = idx;
                            continue;
                        }
                        None => break,
                    }
                }
            } else {
                latest = idx;
                break;
            }
        } else if c == '#' {
            loop {
                match chars.next() {
                    Some((idx, '\n')) => {
                        latest = idx;
                        break;
                    }
                    Some((_, _)) => continue,
                    None => break,
                }
            }
        } else {
            break;
        }
    }
    contents.get(latest..).unwrap()
}
#[cfg(test)]
mod tests {
    use solite_stdlib::BUILTIN_FUNCTIONS;

    use super::*;
    use crate::{dot::DotCommand, sqlite::Connection};

    #[test]
    fn test_advance_through_ignorable() {
        assert_eq!(advance_through_ignorable("4"), "4");
        assert_eq!(advance_through_ignorable("    4"), "4");

        assert_eq!(advance_through_ignorable("--\n4"), "4");
        assert_eq!(advance_through_ignorable("-- skip me\n4"), "4");
        assert_eq!(
            advance_through_ignorable("-- skip me\n  -- 2nd line \n 4"),
            "4"
        );
        assert_eq!(advance_through_ignorable("-\n4"), "-\n4");

        assert_eq!(advance_through_ignorable("#\n4"), "4");
        assert_eq!(advance_through_ignorable("# skip me\n4"), "4");
        assert_eq!(advance_through_ignorable("# skip me\n  #\n 4"), "4");
        assert_eq!(advance_through_ignorable("#\n--\n4"), "4");

        assert_eq!(advance_through_ignorable("/**/4"), "4");
        assert_eq!(advance_through_ignorable("/* skip me */ 4"), "4");
        assert_eq!(advance_through_ignorable("/** skip me */ 4"), "4");
        assert_eq!(advance_through_ignorable("/* skip me **/ 4"), "4");
    }

    fn functions_of(db: &Connection) -> Vec<String> {
        let stmt = db
            .prepare("select distinct name from pragma_function_list order by 1")
            .unwrap()
            .1
            .unwrap();
        let mut fns = vec![];
        while let Ok(Some(row)) = stmt.next() {
            fns.push(row.get(0).unwrap().as_str().to_string());
        }
        fns
    }
    fn version_functions_of(db: &Connection) -> Vec<String> {
        let stmt = db
            .prepare("select distinct name from pragma_function_list where name like '%_version' order by 1")
            .unwrap()
            .1
            .unwrap();
        let mut fns = vec![];
        while let Ok(Some(row)) = stmt.next() {
            fns.push(row.get(0).unwrap().as_str().to_string());
        }
        let mut sql = String::new();
        sql += "select ";
        sql += &fns
            .iter()
            .map(|v| format!("{v}()"))
            .collect::<Vec<String>>()
            .join(", ");
        let stmt = db.prepare(&sql).unwrap().1.unwrap();
        let row = stmt.next().unwrap().unwrap();
        let vers = row.iter().map(|v| v.as_str().to_string());
        assert_eq!(fns.len(), vers.len());
        //fns.iter().zip(vers).collect()
        fns.iter()
            .zip(vers)
            .map(|(func, version)| format!("{func} == {version}"))
            .collect()
    }
    fn modules_of(db: &Connection) -> Vec<String> {
        let stmt = db
            .prepare("select distinct name from pragma_module_list order by 1")
            .unwrap()
            .1
            .unwrap();
        let mut mods = vec![];
        while let Ok(Some(row)) = stmt.next() {
            mods.push(row.first().unwrap().as_str().to_string());
        }
        mods
    }

    #[test]
    fn core_basic() {
        let runtime = Runtime::new(None);
        let stmt = runtime
            .connection
            .prepare("select sqlite_version();")
            .unwrap()
            .1
            .unwrap();
        assert_eq!(
            stmt.next().unwrap().unwrap().get(0).unwrap().as_str(),
            "3.49.1"
        );
        insta::assert_yaml_snapshot!(functions_of(&runtime.connection));
        insta::assert_yaml_snapshot!(modules_of(&runtime.connection));
        insta::assert_yaml_snapshot!(version_functions_of(&runtime.connection));
        insta::assert_yaml_snapshot!(BUILTIN_FUNCTIONS);
    }
    #[test]
    fn core_stack() {
        let mut runtime = Runtime::new(None);
        runtime.enqueue(
            "[input]",
            "create table t(a);
                insert into t select 1;
                insert into t select 2; ",
            BlockSource::File(PathBuf::new()),
        );
        runtime.execute_to_completion().unwrap();
        let binding = runtime
            .connection
            .prepare("select json_group_array(a) from t")
            .unwrap()
            .1
            .unwrap();
        let binding = binding.next().unwrap().unwrap();
        let x = binding.get(0).unwrap().as_str();
        assert_eq!(x, "[1,2]");

        runtime.enqueue(
            "[input]",
            ".print yo!
.print ok
select 4;",
            BlockSource::Repl,
        );

        match runtime.next_step().unwrap().unwrap().result {
            StepResult::DotCommand(cmd) => {
                assert_eq!(
                    cmd,
                    DotCommand::Print(dot::PrintCommand {
                        message: "yo!".to_owned()
                    })
                )
            }
            _ => panic!("fail"),
        };
        match runtime.next_step().unwrap().unwrap().result {
            StepResult::DotCommand(cmd) => {
                assert_eq!(
                    cmd,
                    DotCommand::Print(dot::PrintCommand {
                        message: "ok".to_owned()
                    })
                )
            }
            _ => panic!("fail"),
        };
        match runtime.next_step().unwrap().unwrap().result {
            StepResult::SqlStatement{stmt, ..} => {
                let row = stmt.next().unwrap().unwrap();
                assert_eq!(row.first().unwrap().as_str(), "4");
            }
            _ => panic!("fail"),
        };
        assert!(runtime.next_step().unwrap().is_none())
        /*runtime
        .execute_to_completion(
            "[input]",
            "select * from \"sample.csv\"",
            BlockSource::File(PathBuf::new()),
        )
        .unwrap();*/
    }
}
