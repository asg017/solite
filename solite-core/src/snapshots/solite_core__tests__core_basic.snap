---
source: solite-core/src/lib.rs
expression: functions_of(&runtime.connection)
snapshot_kind: text
---
- "->"
- "->>"
- abs
- acos
- acosh
- asin
- asinh
- atan
- atan2
- atanh
- avg
- base64
- bm25
- ceil
- ceiling
- changes
- char
- clipboard_set
- coalesce
- concat
- concat_ws
- cos
- cosh
- count
- csv_at
- csv_line_number
- csv_path
- cume_dist
- current_date
- current_time
- current_timestamp
- date
- datetime
- decimal
- decimal_add
- decimal_cmp
- decimal_exp
- decimal_mul
- decimal_pow2
- decimal_sub
- decimal_sum
- degrees
- dense_rank
- editdist3
- exp
- first_value
- floor
- format
- fts5
- fts5_get_locale
- fts5_locale
- fts5_source_id
- geopoly_area
- geopoly_bbox
- geopoly_blob
- geopoly_ccw
- geopoly_contains_point
- geopoly_debug
- geopoly_group_bbox
- geopoly_json
- geopoly_overlap
- geopoly_regular
- geopoly_svg
- geopoly_within
- geopoly_xform
- glob
- group_concat
- hex
- highlight
- http_debug
- http_get_body
- http_request
- http_version
- ieee754
- ieee754_exponent
- ieee754_from_blob
- ieee754_inc
- ieee754_mantissa
- ieee754_to_blob
- ifnull
- iif
- instr
- json
- json_array
- json_array_length
- json_error_position
- json_extract
- json_group_array
- json_group_object
- json_insert
- json_object
- json_patch
- json_pretty
- json_quote
- json_remove
- json_replace
- json_set
- json_type
- json_valid
- jsonb
- jsonb_array
- jsonb_extract
- jsonb_group_array
- jsonb_group_object
- jsonb_insert
- jsonb_object
- jsonb_patch
- jsonb_remove
- jsonb_replace
- jsonb_set
- julianday
- lag
- last_insert_rowid
- last_value
- lead
- length
- like
- likelihood
- likely
- lines_debug
- lines_version
- ln
- load_extension
- log
- log10
- log2
- lower
- lsmode
- ltrim
- match
- max
- min
- mod
- nth_value
- ntile
- nullif
- octet_length
- path_at
- path_debug
- path_version
- percent_rank
- pi
- pow
- power
- printf
- quote
- radians
- random
- randomblob
- rank
- readfile
- regex
- regex_capture
- regex_debug
- regex_find
- regex_find_at
- regex_print
- regex_replace
- regex_replace_all
- regex_valid
- regex_version
- regexp
- regexset
- regexset_is_match
- regexset_print
- replace
- round
- row_number
- rtreecheck
- rtreedepth
- rtreenode
- rtrim
- sha1
- sha1_query
- sha1b
- sha3
- sha3_agg
- sha3_query
- sign
- sin
- sinh
- snippet
- solite_stdlib_version
- soundex
- spellfix1_editdist
- spellfix1_phonehash
- spellfix1_scriptcode
- spellfix1_translit
- sqlite_compileoption_get
- sqlite_compileoption_used
- sqlite_log
- sqlite_source_id
- sqlite_version
- sqrt
- strftime
- string_agg
- substr
- substring
- subtype
- sum
- tan
- tanh
- tg_contains
- tg_coveredby
- tg_covers
- tg_debug
- tg_disjoint
- tg_extra_json
- tg_geom
- tg_group_feature_collection_geojson
- tg_group_geometry_collection
- tg_group_multipoint
- tg_intersects
- tg_multipoint
- tg_point
- tg_to_geojson
- tg_to_wkb
- tg_to_wkt
- tg_touches
- tg_type
- tg_valid_geojson
- tg_valid_wkb
- tg_valid_wkt
- tg_version
- tg_within
- time
- timediff
- total
- total_changes
- trim
- trunc
- tsv_at
- tsv_line_number
- tsv_path
- typeof
- ulid
- ulid_bytes
- ulid_datetime
- ulid_debug
- ulid_version
- ulid_with_datetime
- ulid_with_prefix
- unhex
- unicode
- unixepoch
- unlikely
- upper
- url_debug
- url_escape
- url_fragment
- url_host
- url_path
- url_port
- url_query
- url_scheme
- url_valid
- url_version
- usleep
- uuid
- uuid_blob
- uuid_str
- vec_add
- vec_bit
- vec_debug
- vec_distance_cosine
- vec_distance_hamming
- vec_distance_l1
- vec_distance_l2
- vec_f32
- vec_int8
- vec_length
- vec_normalize
- vec_quantize_binary
- vec_quantize_int8
- vec_slice
- vec_sub
- vec_to_json
- vec_type
- vec_version
- writefile
- xsv_at
- xsv_debug
- xsv_line_number
- xsv_path
- xsv_version
- zeroblob
